Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-85.ca-central-1.compute.internal
Sha: 9a215da7d778aa01d40c4166a8bbc34e484f781e
Number of Error(s): 8
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
\terrors: 0
/home/centos/smoketest/PR-14252/HPCC-Platform/roxie/ccd/ccdserver.cpp:2308:17: error: expected \xe2\x80\x98;\xe2\x80\x99 before \xe2\x80\x98row\xe2\x80\x99
/home/centos/smoketest/PR-14252/HPCC-Platform/roxie/ccd/ccdserver.cpp:2341:28: error: return-statement with a value, in function returning void [-fpermissive]
make[2]: *** [roxie/ccd/CMakeFiles/ccd.dir/build.make:115: roxie/ccd/CMakeFiles/ccd.dir/ccdserver.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:16458: roxie/ccd/CMakeFiles/ccd.dir/all] Error 2
make: *** [Makefile:163: all] Error 2
Get ECLWatch build result
ReportTimes.
Finish at: 2020-11-03_17-43-59
time logs started
Prep time   :     9 sec (00:00:09)
Build time  :   439 sec (00:07:19)
Package time:     0 sec (00:00:00)
Install time:     0 sec (00:00:00)
Start time  :     0 sec (00:00:00)
Test time   :     0 sec (00:00:00)
Stop time   :     0 sec (00:00:00)
Summary     :   448 sec (00:07:28)
time logs end

[Build error:](https://gist.github.com/5d84eb07290097575aee3c6b71e0c230)
Build log [Build log](https://gist.github.com/HPCCSmoketest/5d84eb07290097575aee3c6b71e0c230#file-result-20-11-03-17-36-19-log)

--- a/roxie/ccd/ccdserver.cpp
+++ b/roxie/ccd/ccdserver.cpp
@@ -2184,7 +2184,8 @@ protected:
     IEngineRowStream *inputStream;
     IRecordPullerCallback *helper;
     Semaphore started;                      // MORE: GH->RKC I'm pretty sure this can be deleted, since handled by RestartableThread
-    bool groupAtOnce, eof, eog;
+    bool groupAtOnce, eog;
+    std::atomic<bool> eof;
     CriticalSection crit;
 
 public:
@@ -2277,6 +2278,7 @@ public:
             CriticalBlock c(crit); // stop is called on our consumer's thread. We need to take care calling stop for our input to make sure it is not in mid-nextRow etc etc.
             if (inputStream)
                 inputStream->stop();
+            eof = true;
         }
         RestartableThread::join();
     }
@@ -2318,12 +2320,13 @@ public:
     {
         if (eof)
             return false;
-        while (preload)
+        while (preload && !eof)
         {
-            const void * row;
+            const void * row = nullptr;
             {
                 CriticalBlock c(crit); // See comments in stop for why this is needed
-                row = inputStream->nextRow();
+                if (!eof)
+                    row = inputStream->nextRow();
             }
             if (row)
             {
@@ -2352,10 +2355,11 @@ public:
         unsigned rowsDone = 0;
         while (preload && !eof)
         {
-            const void *row;
+            const void *row = nullptr;
             {
                 CriticalBlock c(crit);
-                row = inputStream->nextRow();
+                if (!eof)
+                    row = inputStream->nextRow();
             }
             if (row)
             {

